src/govuk/components/error-summary/error-summary.mjs

```diff
@@ -98,16 +98,20 @@ ErrorSummary.prototype.handleClick = function (event) {
  * NVDA (as tested in 2018.3.2) - without this only the field type is announced
  * (e.g. "Edit, has autocomplete").
  *
- * @param {HTMLElement} $target - Event target
+ * @param {EventTarget | HTMLAnchorElement} $target - Event target
  * @returns {boolean} True if the target was able to be focussed
  */
 ErrorSummary.prototype.focusTarget = function ($target) {
   // If the element that was clicked was not a link, return early
-  if ($target.tagName !== 'A' || $target.href === false) {
+  if ($target.tagName !== 'A' || !('href' in $target)) {
```

Comment:
**suggestion** I've managed to make TypeScript happy with:

1. typing the param as `Element` which is what the code actually expects. The component runs some event delegation where it listens on clicks on anything within its root and only reacts when clicking on links. So `target` can legitimately be any other kind of `Element` than an `HTMLAnchorElement`.
2. then, casting `target` as an `Element` when calling `focusTarget` L81. That lets TypeScript be happy with that

	```js
	if (this.focusTarget(/** @type Element */ (target))) {
	```
3. finally, inside `focusTarget` casting `$target` as an `HTMLAnchorElement` once we've passed the `if` as by that point, we know what this is.

	```js
	var inputId = this.getFragmentFromUrl(/** @type HTMLAnchorElement */ ($target).href)
	```
	
I would argue that all that type wrangling should happen in one place inside the `handleClick` method by moving the initial `if` there. Guess that TypeScript not only highlights typing issue but also leaky separation of concerns 🤔 

```js
ErrorSummary.prototype.handleClick = function (event) {
  var $target = event.target
  // We're delegating the listening to the root of the component
  // but are actually only interested in clicks on links
  if (/** @type Element */ ($target).tagName !== 'A' || !('href' in $target)) {
    if (this.focusTarget(/** @type HTMLAnchorElement */ ($target))) {
     event.preventDefault()
    }
  }
}
```

Reply:
Ahh I'm afraid I don't want to do any type casting as it hides potential issues from the compiler

Good spot though, I just need to flip this one? This works:

```suggestion
  if (!('href' in $target) || $target.tagName !== 'A') {
```

But ideally if we had good browser support we'd go with:

```suggestion
  if (!($target instanceof HTMLAnchorElement)) {
```

Reply:
Good ideas about the separate of concerns though 😊

Reply:
I'm not sure I follow how we can sort this situation without casting. Even with an `instanceof` TypeScript doesn't seem to recognise that the target is an `HTMLAnchorElement` (that would have been neat, though):

<img width="1178" alt="Screenshot 2022-11-21 at 12 41 21" src="https://user-images.githubusercontent.com/396367/203058032-97f36f64-bc6b-44e7-8dd6-12e190aafd3e.png">

And in the end that casting reflects what we do with the code: we get provided a very generic type from `event.target` then checks after checks we narrow down its specialized type to use as it becomes necessary. 

I understand that casting could be a way to hide issues from compilers and should generally be avoided, but I don't think it can be avoided at all times, especially when looking at algorithms that rely on generic types becoming more specialised like event delegation. 

Given we're in control of what's around, it's pretty safe in our situation: 

- we know that his particular `event.target` provides an element as the `addEventListener` is called on a DOMElement
- and after the `if` we know it's an `HTMLAnchorElement` (or `SVGAElement` actually given our check, but that doesn't really matter for what the method does)). 

Even if `event.target` was typed more specifically from `MouseEvent`, we'd still need that second casting once we've checked which element we're manipulating.

Reply:
A tidier way to cast may be:

```js
ErrorSummary.prototype.handleClick = function (event) {
  var $target = /** @type Element */ (event.target)
  // We're delegating the listening to the root of the component
  // but are actually only interested in clicks on links
  if ($target.tagName !== 'A' || !('href' in $target)) {
    if (this.focusTarget(/** @type HTMLAnchorElement */ ($target))) {
     event.preventDefault()
    }
  }
}
```

Reply:
Mind checking this for me @romaricpascal?

I was over-thinking it. Internet Explorer 8 works fine with:

```js
$target instanceof Element
$target instanceof HTMLElement
$target instanceof HTMLAnchorElement
```

See commit https://github.com/alphagov/govuk-frontend/pull/2987/commits/e475f24c98ab43e738fd4581c122f29adc3dc346

Thankfully nothing complicated is needed 😊

Reply:
I might misunderstand, but these checks look like they're actually masking errors, rather than revealing them:

If an object with the wrong type ends up there, that'd be missing a piece of API from the expected type, we won't know. No error will be throwns, things will just fail silently. Whether checks are here and bypass what's inside the `if`, or they're not and an error is thrown, the component won't behave as intedended so we haven't gained any security.

Even if we consider logging in an `else` where console is supported (or using [`console.assert`](https://developer.mozilla.org/en-US/docs/Web/API/console/assert), would TypeScript be happy with it?), that feels like we've just added more code to get to the same result as letting things be and the browser throwing an error, except now it's on our hands to make sure we `if` and log.

Reply:
Yeah 100%

I've used the approach we follow currently

1. Call function
2. Do some checks
3. Return early without output

For example [**`error-summary.mjs`**](https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/components/error-summary/error-summary.mjs#L104) we return `false` after each check
```js
// If the element that was clicked was not a link, return early
if ($target.tagName !== 'A' || $target.href === false) {
  return false
}

var inputId = this.getFragmentFromUrl($target.href)
var $input = document.getElementById(inputId)
if (!$input) {
  return false
}

var $legendOrLabel = this.getAssociatedLegendOrLabel($input)
if (!$legendOrLabel) {
  return false
}
```

@36degrees What do you think?

I'm more than happy to add logging, but primarily this is to teach the compiler what we _really_ meant.

It's pedantic of `tsc` but it's correct, you could pass a plain object in and it'd pass the link check:

```mjs
const $target = {
  tagName: 'A',
  href: ['this is a strange link']
}

if ($target.tagName !== 'A' || $target.href === false) {
  return false
}

// Yep, it must be a link
```

Reply:
> I was over-thinking it. Internet Explorer 8 works fine with:
> 
> ```js
> $target instanceof Element
> $target instanceof HTMLElement
> $target instanceof HTMLAnchorElement
> ```

Just to clarify, `tsc` also recognises this check is happening, and doesn't flag when you call functions that are only present on an `HTMLAnchorElement`?

> If an object with the wrong type ends up there, that'd be missing a piece of API from the expected type, we won't know. No error will be throwns, things will just fail silently.

I'm not 100% sure I'm following, but this is one of a few places in our components where we currently set up event listeners on the `$module` (or some other container) but only actually care about clicks on particular things within it.

So in this example, a click on _anything_ inside the error summary will (by design) result in this function being called, but if it's not a link we bail out early as we just don't care. I would not expect that to result in an error being thrown or anything being logged. Everything is 'working as designed'.

Reply:
> Just to clarify, `tsc` also recognises this check is happening, and doesn't flag when you call functions that are only present on an `HTMLAnchorElement`?

Looks like TypeScript is clever enough to understand that within an `if (variable instanceof X)` or after an `if (!(variable instanceof X)) { return }`, `variable` can be considered of type `X`. 

This is great, but makes me wary of proliferation of `instanceof` checks. In the long run, I'm worried we won't be able to distinguish those that are in to make TypeScript happy from those there for an actual reason in the code.

I'm not sure the error summary illustrates that well, as moving to `instanceof` could be another very valid way of handling the event delegation. But in cases like the radio, we end up with a check that the `$input` is an `HTMLInputElement` right after having explicitely `querySelectorAll` with `input` in the selector:

```js
  var $allInputs = document.querySelectorAll('input[type="radio"][aria-controls]')

  var $clickedInputForm = $clickedInput.form
  var $clickedInputName = $clickedInput.name

  nodeListForEach($allInputs, function ($input) {
    if ($input instanceof HTMLInputElement) {
      var hasSameFormOwner = $input.form === $clickedInputForm
      var hasSameName = $input.name === $clickedInputName

      if (hasSameName && hasSameFormOwner) {
        this.syncConditionalRevealWithInputState($input)
      }
    }
  }.bind(this))
```

Tabs also get some extra checks for stuff we're pretty certain of (because we work on an HTML DOM tree, I don't think an event target node can be nested in anything else than an element, so why would our code explicitely check that):

```js
  if (currentTab.parentNode instanceof Element) {
    var previousTabListItem = currentTab.parentNode.previousElementSibling
    if (previousTabListItem) {
      var previousTab = previousTabListItem.querySelector('.govuk-tabs__tab')
    }

    if (previousTab instanceof HTMLAnchorElement) {
      this.hideTab(currentTab)
      this.showTab(previousTab)
      previousTab.focus()
      this.createHistoryEntry(previousTab)
    }
  }
```

I'd argue typing comments (`/** @type X */ (variable)`) have several advantages over these `if`s:
- they won't be shipped to clients after minification
- they make it explicit that we're type casting, vs muddying`instanceof` which has valid use for generic algorithms like event delegation
- they don't reduce safety given we know our functions operate in a specific environment (HTML DOM tree)
- they'd actually let any error, due to malformed markup for ex, surface in the console for devs and be caught by error reporting tools (Sentry, Bugsnag or whatever the service uses), rather than be swallowed silently never to be recognised

Reply:
>Here's an example that passes the checks and crashes the code:
>https://codepen.io/colinrotherham/pen/NWzLgXL

@romaricpascal Have a look at the CodePen as using `/** @type X */` wouldn't stop the code crashing

It's a nice example though of how types can prevent bugs, harden our code and reduce testing

Don't forget too—when importing `govuk-frontend` from source—that a good minifier can remove all the `instanceof` checks since unused `if/else` branches can be flagged as unreachable code

Reply:
> I'd argue typing comments (`/** @type X */ (variable)`) have several advantages over these `if`s:
> 
> * they won't be shipped to clients after minification
> * they make it explicit that we're type casting, vs muddying`instanceof` which has valid use for generic algorithms like event delegation
> * they don't reduce safety given we know our functions operate in a specific environment (HTML DOM tree)

We'd need to make sure that we used them 'responsibly' – we then need to be able to distinguish between using them when _we know_ that an element can only possibly be of that type, and when _we're assuming_  that an element is going to be of a certain type. It puts the onus back on us to consider every possible way the code can be called.

If a new developer or contributor doesn't understand the different, I think it would be relatively easy for these to start creeping in across the codebase, arguably undoing some of the benefits that we get from using tsc.

I'm not necessarily against it, but I think we'd need some 'rules' about exactly when we use them.

> * they'd actually let any error, due to malformed markup for ex, surface in the console for devs and be caught by error reporting tools (Sentry, Bugsnag or whatever the service uses), rather than be swallowed silently never to be recognised

I think we need to spend some time working out the desired behaviour a little better… when and how do we want the component to throw errors, and how much guarding should the component be doing?

> With `ErrorSummary.prototype.focusTarget()` being "public" it's open for anyone to call. Another organisation might use `OrganisationErrorSummary extends ErrorSummary` and override `.init()`.
> 
> Here's an example that passes the checks and crashes the code: https://codepen.io/colinrotherham/pen/NWzLgXL

Although technically these methods are (more by accident than by design) public, I think we need to be mindful of whether we reasonably expect anyone to actually be calling them from outside of the component. I think in most cases the answer to that is going to be no.

We've certainly removed and renamed methods that are technically 'public' as part of non-breaking changes, on the basis that they're undocumented and that most of them don't make sense outside of the context of the component.

Generally I'm wary of us making changes purely to defend against things that are technically possible but unlikely to actually happen.

Reply:
Agree with everything said 😊

Only need a few tweaks to both help the compiler and harden our checks, like:

### Before
```mjs
if ($target.tagName !== 'A' || $target.href === false) {
```

### After
```mjs
if ($target instanceof HTMLAnchorElement) {
```

Then out the other side comes all the errors/warnings that helped us fix things like:

* https://github.com/alphagov/govuk-frontend/pull/2997

Reply:
Marking this as resolved

I'll keep nice compact `instanceof` checks like this one but move any "just to make the compiler happy" into another PR we can consider separately